Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift

```diff
@@ -1226,7 +1244,13 @@ extension MacroApplication {
     else {
       return .notAMacro
     }
+
     do {
+      guard expandingFreestandingMacros.allSatisfy({ $0 != macro }) else {
+        throw MacroExpansionError.circularExpansion(macro, node)
+      }
+      expandingFreestandingMacros.append(macro)
+
       if let expanded = try expandMacro(macro, node) {
```

Comment:
Instead of removing macros in a number of places, can’t we just remove `macro` from `expandingFreestandingMacros` after the `if`?

Reply:
> Instead of removing macros in a number of places, can’t we just remove `macro` from `expandingFreestandingMacros` after the `if`?

I've tried this approach before only to find out there could be non-intermediate recursive macro expansion hidden inside `expanded`, e.g. `#macro1 -> #macro2 -> #macro1`. Since `#macro1` expands into `#macro2`, if we pop `#macro1` from the stack right after `expandMacro(macro, node)`, by the time `#macro2` is expanded into `#macro1`, the stack won't contain `#macro1`, resulting in infinite expansion.

So we can safely discard a macro name only when we can be sure the macro expansion has fully completed, i.e. the sites I've placed the removal logic. (I might have missed some other sites though)

> We could make `expandingFreestandingMacros` a `Set` to avoid the quadratic complexity and use `!expandingFreestandingMacros.contains(macro)`

The reason I'm using an `Array` instead of a `Set` here is I need to keep track of the insertion order, such that I can safely discard all macro names indexed on or after `macroCount` because they must be inserted during the expansion.

If asymptotic performance is a concern, I can use a `Set` for membership checking alongside an `Array` for keeping track of insertion order.